### PR TITLE
 sys_overlay/process: Implement ppc segmentation flag

### PR DESCRIPTION
**Pull request by elad335**
What ppc segmentation flag does: reserves one segment for fixed allocations by sys_overlay, if it is not set ENOSYS is returned.
Allows Metal Gear Solid 4 to boot.
I was needed to modify my testcases' self files with a hex editor to discover this.

image

### CODE REVIEW

**Comment by Xcedf** [comment]
Is this screen from modified self? cause the game still not goes in menu after installing, shows blackscreen

**Comment by elad335** [comment]
upload log.

**Comment by elad335** [comment]
actually no, disc version has saperate issues, and this screen is not ingame.

**Comment by Xcedf** [comment]
So PSN version works to the screen state?
the log if still needed
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3124935/RPCS3.log.gz)

**Comment by elad335** [comment]
@Xcedf try with lle libvdec.sprx, you can also press x to skip the blackscreen where an intro was supposed to be shown iirc, something similar happens on psn version.

**Comment by Xcedf** [comment]
I know this PR mostly about MGS4 but zcull commit is pretty cool by itself
some games that rely on zcull now has their graphics intact with Zcull Disabled and have fps improvement
GoW3 Demo Zcull ON

Zcull Off

RDR unfortunately framerate improved only in cutscenes not in a gameplay
Zcull On

Zcull OFF

**Comment by salvy** [comment]
I went through my library and only Asura's Wrath its not working for me with this PR. I attached my log, hopefully that helps
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3125486/RPCS3.log.gz)

**Comment by Xcedf** [comment]
Found another little improvement. in Metro Last Light this new ZCull Off even more accurate than ZCull on, it fixes disappearing shadows:
ZCull ON

ZCull Off

**Comment by Xcedf** [comment]
Another little improvement, Zcull off in this PR fixes dissapearing shadows in Metro: Last Light
ZCull On

Zcull OFF

File: rpcs3/Emu/RSX/RSXThread.cpp [hunk 1]
```diff
@@ -2222,7 +2221,7 @@ namespace rsx
 		}
 
 		vm::ptr<CellGcmReportData> result = sink;
-		result->value = value;
+		result->value = UINT32_MAX;
```
Comment (kd-11) [inline rpcs3/Emu/RSX/RSXThread.cpp]: Now we need another option to actually disable ZCULL. It is important for debugging overdraw problems you know. Implement this as a combobox with three options for ZCULL behaviour -> "Default (Hardware)", "Always Pass (-1u)", "Always Fail (0)"

**Comment by kd-11** [comment]
@Xcedf ZCULL off is a hack and a nasty one at that - consider opening a bug report for that metro issue. The same should be done for any issues that only occur when proper rendering is used (WCB, Strict mode, etc), the accurate path must be fixed/always correct.

**Comment by elad335** [comment]
@salvy Does increasing ppu threads count in config.yml to 3 fixes this?

**Comment by MSuih** [comment]
Skate 3 doesn't go ingame with this PR anymore, crashes when loading the world. Seems to pop one of these two error messages:
Access violation reading 0x0 - [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3128680/RPCS3.log.gz)
Verification failed in sys_cond.cpp:284 - [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3128690/RPCS3.log.gz)

**Comment by elad335** [comment]
@MSuih Fixed, @salvy Retest.

**Comment by salvy** [comment]
@elad335 
Re-tested with the latest build and it works now, thanks!

**Comment by jhonyk9b** [comment]
How can i dawnload and ise this new pr mgs4?please(:

File: rpcs3/Emu/Cell/lv2/sys_spu.cpp [hunk 2]
```diff
@@ -172,10 +171,30 @@ error_code sys_spu_image_open(vm::ptr<sys_spu_image> img, vm::cptr<char> path)
...
+	sys_spu.warning("sys_spu_image_open_by_fd(img=*0x%x, fd=%d)", img, fd);
+
+	const auto file = idm::get<lv2_fs_object, lv2_file>(fd);
+
+	if (!file)
+	{
+		return CELL_EBADF;
+	}
+
+	return spu_image_open(img, fmt::format("%s_x%x", file->name.data(), offset));
```
Comment (kd-11) [inline rpcs3/Emu/Cell/lv2/sys_spu.cpp]: Given the fd error check here, why not check for ENOENT in syst_spu_image_open and simplify spu_image_open by passing a file descriptor already pointing to the data?

**Comment by ju2ef1986** [comment]
I'd like to contribute by testing this PR. If you let me know what to look for and how/what to submit I will be more than happy to help.

**Comment by fdemian** [comment]
Hi! 

I can confirm that this changes get Metal Gear Solid 4 [BLES00246] to somewhat work. I get to the Kojima productions logo, then the Konami logo , then the Dolby Digital logo (which takes nearly a minute and a half to show up) and then the screen goes black and it stops displaying things.

File: rpcs3/Emu/Cell/lv2/sys_spu.cpp [hunk 3]
```diff
@@ -159,21 +160,39 @@ error_code _sys_spu_image_get_information(vm::ptr<sys_spu_image> img, vm::ptr<u3
...
 
-	if (!elf_file)
+	if (!file)
 	{
-		sys_spu.error("sys_spu_image_open() error: failed to open %s!", path);
 		return CELL_ENOENT;
 	}
 
-	img->load(elf_file);
+	return spu_image_open(img, std::move(file));
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_spu.cpp]: I'm only using std::move for it to compile due to deleted constructors, the object isn't being used later so I think it's fine.

**Pull request closed without merging**